### PR TITLE
Generate all .o .deps and binaries to generated/

### PR DESCRIPTION
**Pull request by RazvanN7**
I created a directory src/ddmd where I moved all the files in src/, except makefiles. Other than that I modified optabgen.d to generate sources to src/ddmd/backend and I also moved tk.c to backend since it was the only .c file in src/ddmd.

I also modified the makefile to have all generated files (except .c .d and .h files - source code generated files) in a directory called generated/ which is on the root level (same level as src). After creating dmd, the executable is copied to src/dmd so that other tools aren't affected by this PR. 

Now all the other tools can be updated to use the dmd from the generated/ folder whithout causing a major stall. After all the scripts are updated to use generated/dmd/ I will issue a new PR which renames ddmd to dmd and no longer generates dmd in the src folder. This way things will get smoothly in the right direction.

This is a WIP as I only applied the modifications to the posix makefile. 

This is a duplicate of #6489 since it was more time consuming to apply these changes on that branch.

### CODE REVIEW

File: src/posix.mak [hunk 1]
```diff
@@ -22,9 +22,15 @@ INSTALL_DIR=../../install
...
-ROOT=root
+D = ddmd
+
+C=$D/backend
+TK=$D/tk
+ROOT=$D/root
+
+GENERATED = ../generated
+G = $(GENERATED)/$(OS)$(MODEL)
+$(shell mkdir -p $G)
```
Comment (wilzbach) [inline src/posix.mak]: How about following the style used at [Phobos](https://github.com/dlang/phobos/blob/master/posix.mak) and [Druntime](https://github.com/dlang/druntime/blob/master/posix.mak)?

Comment (RazvanN7) [inline src/posix.mak]: I think that there is no good reason to have two unused levels in the hierarchy

**Comment by RazvanN7** [comment]
I don't have a windows setup, so my strategy is to spam the auto-tester. If someone with a windows machine wishes to co-op this PR, be my guest.

File: src/posix.mak (hunk 1 above)
Comment (wilzbach) [inline src/posix.mak]: > I think that there is no good reason to have two unused levels in the hierarchy

Well you can do `debug` and `release` builds at the same time without any issues ... and there are tons of other different build types available for DMD: https://github.com/dlang/dmd/blob/master/src/posix.mak#L160

**Comment by UplinkCoder** [comment]
@RazvanN7 please give me the rational for using the generated folder.
IMO it's easier to work with dmd when the .o files are in the same dir as the sources.

**Comment by wilzbach** [comment]
> please give me the rational for using the generated folder.

https://github.com/dlang/dmd/pull/6489

File: src/posix.mak (hunk 1 above)
Comment (andralex) [inline src/posix.mak]: This is interesting. A few explanations follow.

1. Currently the code does not make the situation any worse than before - it just moves all generated files elsewhere, indiscriminately.

2. The structure `generated/$(OS)$(MODEL) ` (as opposed to Phobos' `generated/$(OS)/$(MODEL)`) was chosen by @RazvanN7 and I after discussing for a while and figuring there's no real need for two separate levels in the dir hierarchy.

3. We could actually simply use `generated/` without the OS or the model, BUT:
  * It is reasonable to generate 32 and 64 binaries on a given machine
  * Even on a given OS, there may be multiple "OS"s - consider generating Windows and Cygwin binaries on Windows. Are there any others? Not sure how important this is, but putting the OS in the dirname helps such situations.

4. Now (this is getting to @wilzbach's point), we have the situation in which changing CFLAGS and DFLAGS affects the generated binaries in many ways, per https://github.com/dlang/dmd/blob/master/src/posix.mak#L160. The "total" solution to that is to canonicalize and place the controlling flags in the dir names, i.e. (after removing `ENABLE`) in alphabetical order: `COVERAGE`, `DEBUG`, `LTO`, `PGO_GENERATE`, `PGO_USE`, `PROFILE`, `PROFILING`, `RELEASE`, `UNITTEST`. Then the dir name of a generated bunch of binaries would be something like `generated/posix64/DEBUG_PROFLE_UNITTEST`. The "final release" version should sit in `generated/$(OS)$(MODEL/` so flags do not affect tool developers. This scheme may be overkill even if technically appropriate. I think @WalterBright and @MartinNowak should chime in here because they'll be the first and most affected.

Comment (MartinNowak) [inline src/posix.mak]: > This is interesting. A few explanations follow.
> 
> 2. The structure generated/$(OS)$(MODEL) (as opposed to Phobos' generated/$(OS)/$(MODEL)) was chosen by @RazvanN7 and I after discussing for a while and figuring there's no real need for two separate levels in the dir hierarchy.

I doesn't cost anything either, and consistency across repos seems more important (in particular for scripting, e.g. release scripts). So let's please just stick with `generated/$OS/$BUILD/$MODEL` as anywhere else.

> Even on a given OS, there may be multiple "OS"s - consider generating Windows and Cygwin binaries on Windows. Are there any others? Not sure how important this is, but putting the OS in the dirname helps such situations.

I also allows to more easily rsync code from other machines/VMs.

> Now (this is getting to @wilzbach's point), we have the situation in which changing CFLAGS and DFLAGS affects the generated binaries in many ways, per https://github.com/dlang/dmd/blob/master/src/posix.mak#L160. The "total" solution to that is to canonicalize and place the controlling flags in the dir names, i.e. (after removing ENABLE) in alphabetical order: COVERAGE, DEBUG, LTO, PGO_GENERATE, PGO_USE, PROFILE, PROFILING, RELEASE, UNITTEST.

Those are hardly maintained and rarely used, let's not overshoot and leave it to maintainers.
Separating debug and relase builds is quite helpful though.
An alternative would be to dump all flags to some file and error on mismatch (requiring a clean step). This is the usual configure/Make approach to changing flags.

Comment (andralex) [inline src/posix.mak]: OK with @MartinNowak, @RazvanN7 can you please execute? Thanks!

**Comment by andralex** [comment]
One more thing: @RazvanN7 mentioned the file `src/checkwhitespace.d` is actually a tool that has no place in the dmd project (it uses phobos, it's also _used_ by phobos and dmd). For this diff @RazvanN7 please leave `checkwhitespace.d` in `src/` to avoid breakage, and let's move it in a later step. Thanks!

**Comment by sprinkle131313** [comment]
With all the changes happening anyways, would there be a possibility of adding druntime/phobos as submodules into dmd? The make files expect a specific file structure for where dmd, druntime, and phobos should be located relative to each other. Having them be submodules in dmd will make building from source easier. It will also know which specific version of druntime/phobos to use, rather than having to checkout the correct version for every repo.

**Comment by MartinNowak** [comment]
> IMO it's easier to work with dmd when the .o files are in the same dir as the sources.

Building out of tree is a common best practice and immediately useful to separate debug and release objects to avoid errors during development.

> With all the changes happening anyways, would there be a possibility of adding druntime/phobos as submodules into dmd?

Let's please not open another can of worms. @CyberShadow has an exterior repo that tracks all repos, https://bitbucket.org/cybershadow/d (used by [Digger](https://github.com/CyberShadow/Digger)).

**Comment by MartinNowak** [comment]
Let's not move `src/checkwhitespace.d → src/ddmd/checkwhitespace.d`, right? Will also fix your tests ;).

**Comment by sprinkle131313** [comment]
@MartinNowak 

This is probably the only time to do so. A lot of the changes that would have to be made are on the same lines that this change is going to cause. Relying on an external tool and external library for building kind of proves my point.

File: src/posix.mak (hunk 1 above)
Comment (CyberShadow) [inline src/posix.mak]: So, what will be the final path of the created `dmd` binary when make is invoked with the default settings?

Comment (RazvanN7) [inline src/posix.mak]: In the end, the path will be generated/dmd, but until then we will keep a copy of dmd in src/ so that we will have a smooth transition

Comment (wilzbach) [inline src/posix.mak]: > the path will be generated/dmd

Didn't we agree on `generated/linux/debug/64/dmd` ?
In the final build step you could create a symlink to the last built binary (e.g `generated/dmd` -> ` `$(abspath generated/linux/debug/64/dmd)`) if you really want to ...

Comment (RazvanN7) [inline src/posix.mak]: That's right. Sorry, I misread the comments

**Comment by RazvanN7** [comment]
The last commit was called "Added $(BUILD) to generated path", but it seems that $(BUILD) was evaluated to the empty string.

**Comment by CyberShadow** [comment]
> The last commit was called "Added $(BUILD) to generated path", but it seems that $(BUILD) was evaluated to the empty string.

I believe the problem was made at the time the commit is created - your shell evaluated `$(BUILD)` to the expansion of a non-existent environment variable. To avoid this, you can use single quotes when specifying the commit message on the command line, or don't use `-m` and type the commit message in the invoked editor instead.

**Comment by RazvanN7** [comment]
@CyberShadow thanks! I should have known that this would happen.

**Comment by andralex** [comment]
Looking green! Well. almost. What's the failure with DAutoTest? Thanks! cc @CyberShadow maybe you could help.

**Comment by CyberShadow** [comment]
```
make[2]: *** No rule to make target 'project.ddoc', needed by '/dev/shm/dtest/work/repo/dlang.org/web/dmd-prerelease/.generated'.  Stop.
```

Is it the DMD source code DDoc stuff? Not really familiar with that.

**Comment by wilzbach** [comment]
> Is it the DMD source code DDoc stuff? Not really familiar with that.

@RazvanN7: Have a look at this PR (https://github.com/dlang/dlang.org/pull/1581/files) for an idea of the relevant bits on dlang.org. It's basically `make -f posix.mak html` within the `src` directory.

**Comment by andralex** [comment]
Cool! So @wilzbach applying the label auto-merge-squash is the procedure now?

File: src/posix.mak [hunk 2]
```diff
@@ -488,13 +513,13 @@ zip:
 
 #############################
 
-ifneq ($(DOCDIR),)
+#ifneq ($(DOCDIR),)
```
Comment (CyberShadow) [inline src/posix.mak]: Why this change, BTW?

I guess the `ifneq` was supposed to protect accidentally writing to the filesystem root... but if you think the change is not needed, I'd say to just remove it. Though, this change looks like it might have been accidentally committed.

File: src/posix.mak [hunk 3]
```diff
@@ -310,36 +332,39 @@ TK_SRC = \
...
+$G/backend.a: $(G_OBJS)
+	$(AR) rcs $@ $(G_OBJS)
 
-dmd_frontend: $(FRONT_SRCS) gluelayer.d $(ROOT_SRCS) newdelete.o $(STRING_IMPORT_FILES) $(HOST_DMD_PATH)
-	CC=$(HOST_CXX) $(HOST_DMD_RUN) -of$@ $(MODEL_FLAG) -vtls -J. -J../res -L-lstdc++ $(DFLAGS) $(filter-out $(STRING_IMPORT_FILES) $(HOST_DMD_PATH),$^) -version=NoBackend
+$G/dmd_frontend: $(FRONT_SRCS) $D/gluelayer.d $(ROOT_SRCS) $G/newdelete.o $(STRING_IMPORT_FILES) $(HOST_DMD_PATH)
+	CC=$(HOST_CXX) $(HOST_DMD_RUN) -of$@ $(MODEL_FLAG) -vtls -J$G -J../res -L-lstdc++ $(DFLAGS) $(filter-out $(STRING_IMPORT_FILES) $(HOST_DMD_PATH),$^) -version=NoBackend
+
+dmd: $G/dmd
+	cp $< .
```
Comment (CyberShadow) [inline src/posix.mak]: `ln` should be a bit more efficient FWIW.

**Comment by wilzbach** [comment]
> Cool! So @wilzbach applying the label auto-merge-squash is the procedure now?

It's not "official" yet as auto-labelled PRs aren't prioritized on the auto-tester (see e.g. https://github.com/dlang-bots/dlang-bot/pull/50), but it will still work and do the squashed merge.
However, I think @RazvanN7 has deliberately made two commits to separate the renaming from the Makefile adjustments, so imho there's no need to do squashing here...

**Comment by CyberShadow** [comment]
> However, I think @RazvanN7 has deliberately made two commits to separate the renaming from the Makefile adjustments, so imho there's no need to do squashing here...

No, the first commit moves the file and the second one moves it back. So a squash is fine in this case.

**Comment by andralex** [comment]
way we go...

**Comment by CyberShadow** [comment]
> So, what will be the final path of the created `dmd` binary when make is invoked with the default settings?

Question still unanswered in light of the latest changes :) Need to know what to update DIgger to use, so that we can delete `dmd/src`..

**Comment by wilzbach** [comment]
> Question still unanswered in light of the latest changes :) Need to know what to update DIgger to use, so that we can delete dmd/src..

AFAIK it's supposed to be `generated/linux/release/64/dmd` (?)

**Comment by RazvanN7** [comment]
@CyberShadow : @wilzbach  is right. that will be the path, same as in the phobos makefile

**Comment by CyberShadow** [comment]
Hmm, dlang.org autodeploy build failed:

```
digger: Running: 'make' '-f' 'posix.mak' 'MODEL=64' 'HOST_DC='
digger: PATH=/bin:/usr/bin
no cpu specified, assuming X86
posix.mak:81: *** 'dmd' not found, get a D compiler or make AUTO_BOOTSTRAP=1.  Stop.
```

**Comment by CyberShadow** [comment]
Something changed in this PR which for some reason kept working on the documentation tester but didn't work on the dlang.org autodeployer, even though they use nearly the same environment and build commands.

**Comment by CyberShadow** [comment]
Found the bug: `if (buildPath(sourceDir, "src", "idgen.d").exists)`

Odd that it worked on DAutoTest :)

**Comment by CyberShadow** [comment]
Fixed.

**Comment by MartinNowak** [comment]
`.gitignore` still needs to be updated

**Comment by MartinNowak** [comment]
Seems like this broken the build script @RazvanN7, could you please have a look.
http://nightlies.dlang.org/dmd-master-2017-03-01/build.html
```
Running: make MODEL=32 DMD=..\dmd\src\dmd HOST_DC=C:\Users\vagrant\old-dmd\dmd2\windows\bin\dmd.exe RELEASE=1 LATEST=master -f win32.mak dmd
if not exist "..\generated\Windows_NT32" mkdir ..\generated\Windows_NT32
make -fwin32.mak C=ddmd\backend TK=ddmd\tk ROOT=ddmd\root MAKE="make" HOST_DC="C:\Users\vagrant\old-dmd\dmd2\windows\bin\dmd.exe" DMODEL= CC="dmc" LIB="lib" OBJ_MSVC="" "OPT=-o" "DEBUG=" "DDEBUG=" "DOPT=-O -release -inline" "LFLAGS=-L/delexe/la" ..\generated\Windows_NT32\dmd.exe
Error: don't know how to make '..\generated\Windows_NT32\dmd.exe'
```

**Comment by CyberShadow** [comment]
@MartinNowak Since so many things seem to have broken nightly builds recently, is there a possibility of adding it as a CI? (Bonus: you can download a version of D that includes a certain PR to test it out)

**Comment by CyberShadow** [comment]
(Alternatively perhaps the nightlies could be built using Digger, which is CI'd through DAutoTest.)

**Comment by RazvanN7** [comment]
@MartinNowak it seems that the rule to make the target ..\generated\Windows...\dmd.exe doesn't exist in the makefile. That is odd, since the rule definitely exists in the makefile that I commited. The link you provided doesn;t output this error. Could you give me some more information?

**Comment by MartinNowak** [comment]
The cost for a real run is prohibitive atm. (too many OSes/VMs).

**Comment by MartinNowak** [comment]
http://nightlies.dlang.org/dmd-master-2017-03-02/build.html, wrong date, my mistake.

**Comment by CyberShadow** [comment]
> The cost for a real run is prohibitive atm. (too many OSes/VMs).

How about making it on-demand, or just whenever a makefile changes?

**Comment by MartinNowak** [comment]
Yes, we should work on replacing the create_dmd_release script and CI integration (heuristic sounds like a good idea) in the longer run, but I'm busy w/ other things atm.
https://trello.com/c/uDoxVgpc/110-ci-for-nightly-builds

**Comment by MartinNowak** [comment]
See https://github.com/dlang/dmd/pull/6585.

**Comment by wilzbach** [comment]
Eww discovered a couple of more issues with the "new" Makefile:

https://github.com/dlang/dmd/pull/6586
